service/frontend/operator_handler.go

```diff
@@ -230,68 +323,151 @@ func (h *OperatorHandlerImpl) RemoveSearchAttributes(ctx context.Context, reques
+		return serviceerror.NewUnavailable(fmt.Sprintf(errUnableToSaveSearchAttributesMessage, err))
+	}
+	return nil
+}
+
+func (h *OperatorHandlerImpl) removeSearchAttributesSQL(
+	ctx context.Context,
+	request *operatorservice.RemoveSearchAttributesRequest,
+) error {
+	client, err := h.clientFactory.NewLocalFrontendClientWithTimeout(
+		frontend.DefaultTimeout,
+		frontend.DefaultLongPollTimeout,
+	)
+	if err != nil {
+		return serviceerror.NewUnavailable(fmt.Sprintf(errUnableToCreateFrontendClientMessage, err))
 	}
 
-	return &operatorservice.RemoveSearchAttributesResponse{}, nil
+	nsName := request.GetNamespace()
+	if nsName == "" {
+		return errNamespaceNotSet
+	}
+	ns, err := h.namespaceRegistry.GetNamespace(namespace.Name(nsName))
+	if err != nil {
+		return serviceerror.NewUnavailable(fmt.Sprintf(errUnableToGetNamespaceInfoMessage, nsName))
+	}
+
+	mapper := ns.CustomSearchAttributesMapper()
+	fieldToAliasMap := maps.Clone(mapper.FieldToAliasMap())
+	for _, saName := range request.GetSearchAttributes() {
+		fieldName, err := mapper.GetFieldName(saName, nsName)
+		if err != nil {
+			return serviceerror.NewNotFound(fmt.Sprintf(errSearchAttributeDoesntExistMessage, saName))
+		}
+		delete(fieldToAliasMap, fieldName)
+	}
+
+	_, err = client.UpdateNamespace(ctx, &workflowservice.UpdateNamespaceRequest{
+		Namespace: nsName,
+		Config: &namespacepb.NamespaceConfig{
+			CustomSearchAttributeAliases: fieldToAliasMap,
+		},
+	})
+	return err
 }
 
-func (h *OperatorHandlerImpl) ListSearchAttributes(ctx context.Context, request *operatorservice.ListSearchAttributesRequest) (_ *operatorservice.ListSearchAttributesResponse, retError error) {
+func (h *OperatorHandlerImpl) ListSearchAttributes(
+	ctx context.Context,
+	request *operatorservice.ListSearchAttributesRequest,
+) (_ *operatorservice.ListSearchAttributesResponse, retError error) {
 	defer log.CapturePanic(h.logger, &retError)
 
 	if request == nil {
 		return nil, errRequestNotSet
 	}
 
 	indexName := h.visibilityMgr.GetIndexName()
-
-	var lastErr error
-	var esMapping map[string]string = nil
-	if h.esClient != nil {
-		esMapping, lastErr = h.esClient.GetMapping(ctx, indexName)
-		if lastErr != nil {
-			lastErr = serviceerror.NewUnavailable(fmt.Sprintf("unable to get mapping from Elasticsearch: %v", lastErr))
-		}
-	}
-
 	searchAttributes, err := h.saProvider.GetSearchAttributes(indexName, true)
 	if err != nil {
-		lastErr = serviceerror.NewUnavailable(fmt.Sprintf("unable to read custom search attributes: %v", err))
-	}
-
-	if lastErr != nil {
-		return nil, lastErr
+		return nil, serviceerror.NewUnavailable(
+			fmt.Sprintf("unable to read custom search attributes: %v", err),
+		)
+	}
+
+	var customSearchAttributes map[string]enumspb.IndexedValueType
+	var storageSchema map[string]string
+	if h.visibilityMgr.GetName() == elasticsearch.PersistenceName {
+		customSearchAttributes = searchAttributes.Custom()
```

Comment:
I think it worth to have `listSearchAttributesElasticsearch` and `listSearchAttributesSQL` here too.